service/tablet_allocator.cc

```diff
@@ -12,12 +12,405 @@
+        ///      3.4. Evaluate collocation constraints for tablet replicas, if they pass:
+        ///           3.4.1 Pick the least-loaded shard on the target
+        ///           3.4.2 Generate a migration for the candidate tablet from the source shard to the target shard
+        ///
+        /// Even though the algorithm focuses on a single target, the fact the the produced plan is just an increment
+        /// means that many under-loaded nodes can be driven forward to balance concurrently because the load balancer
+        /// will alternate between them across make_plan() calls.
+        ///
+        /// The cost of make_plan() is relatively heavy in terms of preparing data structures, so the current
+        /// implementation is not efficient if the scheduler would like to call make_plan() multiple times
+        /// to parallelize execution. This will be addressed in the future by keeping the data structures
+        /// valid across calls and only recalculating them when starting a new round with a new token metadata version.
+        ///
+        /// We can adjust the algorithm to work with multiple targets at a time. We should keep the candidate exclusion set
+        /// per-target.
+
+        lblogger.info("Examining DC {}", dc);
+
+        const locator::topology& topo = _tm->get_topology();
+
+        // Select subset of nodes to balance.
+
+        std::unordered_map<host_id, node_load> nodes;
+        topo.for_each_node([&] (auto node_ptr) {
+            if (node_ptr->get_state() == locator::node::state::normal && node_ptr->dc_rack().dc == dc) {
+                node_load& load = nodes[node_ptr->host_id()];
+                load.tablet_count = 0;
+                load.shard_count = node_ptr->get_shard_count();
+                load.shards.resize(load.shard_count);
+                if (!load.shard_count) {
+                    throw std::runtime_error(format("Shard count of {} not found in topology", node_ptr->host_id()));
+                }
+            }
+        });
+
+        // Compute tablet load on nodes.
+
+        for (auto&& [table, tmap] : _tm->tablets().all_tables()) {
+            auto tid = tmap.first_tablet();
+            for (const tablet_info& ti : tmap.tablets()) {
+                co_await coroutine::maybe_yield();
+                for (auto&& replica : ti.replicas) {
+                    if (nodes.contains(replica.host)) {
+                        nodes[replica.host].tablet_count += 1;
+                        // This invariant is assumed later.
+                        if (replica.shard >= nodes[replica.host].shard_count) {
+                            auto gtid = global_tablet_id{table, tid};
+                            throw std::runtime_error(format("Tablet {} replica {} targets non-existent shard", gtid, replica));
+                        }
+                    }
+                }
+                tid = *tmap.next_tablet(tid);
+            }
+        }
```

Comment:
We should keep the structure updated based on events, but this is good enough to start.